ddtrace/internal/processor.py

```diff
@@ -0,0 +1,31 @@
+from typing import List
+
+from ddtrace import Span
+from ddtrace.filters import TraceFilter
+from ddtrace.internal.writer import TraceWriter
+from ddtrace.vendor import attr
+
+from .logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceProcessor(object):
+    writer = attr.ib(type=TraceWriter, init=True)
+    _filters = attr.ib(type=List[TraceFilter], init=True)
+
+    def process(self, trace):
+        # type: (List[Span]) -> None
+        for filtr in self._filters:
+            try:
+                trace = filtr.process_trace(trace)
+            except Exception:
+                log.error("error applying filter %r to traces", filtr, exc_info=True)
+            else:
+                if trace is None:
+                    log.debug("dropping trace due to filter %r", filter)
+                    return
+
+        self.writer.write(trace)
```

Comment:
I think the API should return the `trace` and not use a `writer` here.
That'd make the processor reusable. It's then up to the caller to write the trace.

Reply:
Great idea!